### PR TITLE
chore: moved optimism_receiptfn and its tests

### PR DESCRIPTION
**Pull request by Redidacove**
Issue
Solves  [#10514](https://github.com/paradigmxyz/reth/issues/10514)

Moved function and its tests to  optimism/primitives

### CODE REVIEW

File: crates/optimism/primitives/src/proof.rs [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+                r
+            })
+            .collect::<Vec<_>>();
+
+        return ordered_trie_root_with_encoder(receipts.as_slice(), |r, buf| {
+            r.encode_inner(buf, false)
+        })
+    }
+
+    ordered_trie_root_with_encoder(receipts, |r, buf| r.encode_inner(buf, false))
```
Comment (onbjerg) [inline crates/optimism/primitives/src/proof.rs:35]: this import is missing, please ensure this compiles

**Review by onbjerg — CHANGES_REQUESTED** [review]

File: crates/optimism/primitives/src/proof.rs [hunk 2]
```diff
@@ -0,0 +1,335 @@
+use crate::{
+     Receipt, ReceiptWithBloom,B256,
+};
+
+/// Calculates the receipt root for a header.
+// #[cfg(feature = "optimism")]
+pub fn calculate_receipt_root_optimism(
```
Comment (onbjerg) [inline crates/optimism/primitives/src/proof.rs:7]: please ensure this is actually used where we currently use it (i.e. fix the import paths elsewhere), see e.g. https://github.com/search?q=repo%3Aparadigmxyz%2Freth%20calculate_receipt_root_optimism&type=code

File: crates/optimism/primitives/src/proof.rs [hunk 3]
```diff
@@ -0,0 +1,335 @@
+use crate::{
+     Receipt, ReceiptWithBloom,B256,
+};
+
+/// Calculates the receipt root for a header.
+// #[cfg(feature = "optimism")]
```
Comment (onbjerg) [inline crates/optimism/primitives/src/proof.rs:6]: rm this

**Review by onbjerg — CHANGES_REQUESTED** [review]

**Comment by onbjerg** [comment]
`cargo check --bin op-reth --features optimism` should do

File: crates/blockchain-tree/src/blockchain_tree.rs [hunk 4]
```diff
@@ -1601,7 +1600,6 @@ mod tests {
             #[cfg(not(feature = "optimism"))]
             let receipts_root = calculate_receipt_root(&receipts);
 
-            #[cfg(feature = "optimism")]
```
Comment (onbjerg) [inline crates/blockchain-tree/src/blockchain_tree.rs:1604]: this is incorrect, please remove this. it should only happen when the optimism feature is enabled.

File: crates/blockchain-tree/src/blockchain_tree.rs [hunk 5]
```diff
@@ -1388,8 +1388,7 @@ mod tests {
     use reth_evm_ethereum::execute::EthExecutorProvider;
     #[cfg(not(feature = "optimism"))]
     use reth_primitives::proofs::calculate_receipt_root;
-    #[cfg(feature = "optimism")]
-    use reth_primitives::proofs::calculate_receipt_root_optimism;
+    use reth_optimism_primitives::proof::calculate_receipt_root_optimism;
```
Comment (onbjerg) [inline crates/blockchain-tree/src/blockchain_tree.rs:1391]: same as below

**Review by onbjerg — CHANGES_REQUESTED** [review]

File: crates/optimism/primitives/src/proof.rs [hunk 6]
```diff
@@ -0,0 +1,330 @@
+use reth_primitives::{
```
Comment (onbjerg) [inline crates/optimism/primitives/src/proof.rs:1]: we need module-level docs here

**Review by onbjerg — CHANGES_REQUESTED** [review]

File: crates/optimism/primitives/src/proof.rs (hunk 6 above)
Comment (Redidacove) [reply crates/optimism/primitives/src/proof.rs:1]: Didn't understood what you are suggesting here what do you mean by module level docs ?

File: crates/optimism/primitives/src/proof.rs
Comment (mattsse) [inline crates/optimism/primitives/src/proof.rs:1]: this does not really work here because of feature issues.

let's move this to consensus crate instead

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/optimism/primitives/src/proof.rs
Comment (Redidacove) [reply crates/optimism/primitives/src/proof.rs:1]: Do you have any package in mind

**Comment by Redidacove** [comment]
moved it to consensus crate @mattsse

File: crates/optimism/consensus/src/proof.rs [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+
+        return ordered_trie_root_with_encoder(receipts.as_slice(), |r, buf| {
+            r.encode_inner(buf, false)
+        })
+    }
+
+    ordered_trie_root_with_encoder(receipts, |r, buf| r.encode_inner(buf, false))
+}
+
+#[cfg(feature = "optimism")]
```
Comment (emhane) [inline crates/optimism/consensus/src/proof.rs:41]:
```suggestion
```

this feature is applied at crate level in _lib.rs_, so no need to put it in the rest of the code in this crate

File: crates/blockchain-tree/src/blockchain_tree.rs [hunk 8]
```diff
@@ -1600,7 +1600,6 @@ mod tests {
 
             #[cfg(not(feature = "optimism"))]
             let receipts_root = calculate_receipt_root(&receipts);
-
             #[cfg(feature = "optimism")]
```
Comment (emhane) [inline crates/blockchain-tree/src/blockchain_tree.rs:1603]: copy this whole test and move it to `reth-optimism-consensus`. we don't want the dep `reth-blockchain-tree` -> `reth-optimism-consensus.`

alternatively, add a test utils module to `reth-blockchain-tree`, move the test body into a new function in that module and make it generic over a type that can calculate the receipts root. that way we avoid redundant code. is also fine to open an issue for that, and just copy paste the test in op crate for now.

**Review by emhane — CHANGES_REQUESTED** [review]

File: crates/blockchain-tree/src/blockchain_tree.rs (hunk 8 above)
Comment (Redidacove) [reply crates/blockchain-tree/src/blockchain_tree.rs:1603]: Which test are you talking about like the while tests suite inside mod tests{}

Comment (emhane) [reply crates/blockchain-tree/src/blockchain_tree.rs:1603]: the `consecutive_reorgs` test

Comment (Redidacove) [reply crates/blockchain-tree/src/blockchain_tree.rs:1603]: ok

Comment (Redidacove) [reply crates/blockchain-tree/src/blockchain_tree.rs]: Did that can you check

**Review by emhane — CHANGES_REQUESTED** [review]
almost there, the test needed to be copied, not moved. the test must run once for non-op once for op

**Comment by Redidacove** [comment]
> almost there, the test needed to be copied, not moved. the test must run once for non-op once for op

@emhane wdym by run once for non-op and once for op ?

**Review by emhane — APPROVED** [review]

**Review by emhane — COMMENTED** [review]
almost there, only lint fixes remaining

**Comment by Redidacove** [comment]
> almost there, only lint fixes remaining

any command to fix

**Review by emhane — COMMENTED** [review]
I think merging main solves those unrelated lint issues

**Review by onbjerg — DISMISSED** [review]
these are not lint fixes, they are missing imports. please import the missing items to ensure this compiles. the error log shows you what to import from where

the imports are only needed in the test, so add them in there:)

**Review by mattsse — APPROVED** [review]
I've removed all the blockchaintree deps from consensus, we deprecate this crate soon anyway

**Pull request merged**